openedx-unsupported/edx-analytics-pipeline
# Store weekly problem history for individual users in Analytics DB

This PR makes problem data for individual users available in the Analytics MySQL DB. It is a first step towards displaying information about a user's interaction with problems belonging to a course in the Insights dashboard.

Related PRs:
- https://github.com/edx/edx-analytics-data-api/pull/81
- https://github.com/edx/edx-analytics-data-api-client/pull/22

These update the data API to enable access to the data from the Insights dashboard.
- https://github.com/edx/edx-analytics-dashboard/pull/343

This dashboard PR adds a page that displays a summary of the data to the user.

**Changes**

Adds three tasks:
- Map/Reduce task for extracting relevant information from event tracking logs
- Task for storing output from map/reduce task in Hive
- Task for transferring data stored in Hive to MySQL DB

**Test instructions: OpenCraft devstack**
1. Setup an [analytics devstack](https://github.com/open-craft/edx-analytics-devstack). 
2. Modify your local LMS devstack to work with the new analytics devstack; see [this section](https://github.com/open-craft/edx-analytics-devstack#lms-setup) of the analytics devstack README. Make sure not to omit step 5 (uploading tracking logs to the analytics devstack's HDFS store).
3. Login to the analytics devstack as the `analytics` user.
4. Run the command:
   `
   launch-task UserProblemHistoryToMySQLTask --local-scheduler --interval=2015-08
   `
5. Wait for it to complete. To check if it worked, run `mysql -u root analytics --execute="SELECT * FROM user_problem_weekly_data;"`. You should get a result set that looks similar to this:

```
+----+-------------+-----------------------+---------+----------------------------------------------------------+--------------+-------------+---------------------+
| id | week_ending | course_id             | user_id | problem_id                                               | num_attempts | final_score | created             |
+----+-------------+-----------------------+---------+----------------------------------------------------------+--------------+-------------+---------------------+
|  1 | 2015-08-17  | edX/DemoX/Demo_Course |       4 | i4x://edX/DemoX/problem/Sample_Algebraic_Problem         |            1 | 1/1         | 2015-08-19 09:08:51 |
|  2 | 2015-08-10  | edX/DemoX/Demo_Course |       4 | i4x://edX/DemoX/problem/0d759dee4f9d459c8956136dbde55f02 |            4 | 0/1         | 2015-08-19 09:08:51 |
+----+-------------+-----------------------+---------+----------------------------------------------------------+--------------+-------------+---------------------+
```

**Test instructions: Combined devstack**
1. Set up a [combined devstack](https://gist.github.com/bradenmacdonald/8b3b9736bf2b5ec44bc0).
2. Clear out any existing pipeline repo: `vagrant ssh` then `sudo rm -r /var/lib/analytics-tasks/devstack/repo` (This is necessary since the ansible scripts won't force the use of a different upstream repo during the next step)
3. On your host, in the edx-analytics-pipeline folder (with the venv loaded), run these commands:
   
   ```
   export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
   remote-task --vagrant-path ../ --remote-name devstack --override-config ${PWD}/config/devstack.cfg --wheel-url $WHEEL_URL --repo https://github.com/open-craft/edx-analytics-pipeline.git --branch user-problem-data-mysql --wait UserProblemHistoryToMySQLTask --local-scheduler --interval 2015-08
   ```
4. Wait for it to complete, then inside the devstack, run `mysql -u root reports --execute="SELECT * FROM user_problem_weekly_data;"`. You should get a result set that looks similar to this:

```
+----+-------------+-----------------------+---------+----------------------------------------------------------+--------------+-------------+---------------------+
| id | week_ending | course_id             | user_id | problem_id                                               | num_attempts | final_score | created             |
+----+-------------+-----------------------+---------+----------------------------------------------------------+--------------+-------------+---------------------+
|  1 | 2015-08-17  | edX/DemoX/Demo_Course |       4 | i4x://edX/DemoX/problem/Sample_Algebraic_Problem         |            1 | 1/1         | 2015-08-19 09:08:51 |
|  2 | 2015-08-10  | edX/DemoX/Demo_Course |       4 | i4x://edX/DemoX/problem/0d759dee4f9d459c8956136dbde55f02 |            4 | 0/1         | 2015-08-19 09:08:51 |
+----+-------------+-----------------------+---------+----------------------------------------------------------+--------------+-------------+---------------------+
```

**Partner information**

Third-party open edX instance, not an edX partner.
